builder/dispatchers.go

```diff
@@ -257,15 +258,63 @@ func run(b *Builder, args []string, attributes map[string]bool, original string)
 		return err
 	}
 
+	// stash the cmd
 	cmd := b.Config.Cmd
-	// set Cmd manually, this is special case only for Dockerfiles
-	b.Config.Cmd = config.Cmd
 	runconfig.Merge(b.Config, config)
+	// stash the config environment
+	env := b.Config.Env
 
 	defer func(cmd *runconfig.Command) { b.Config.Cmd = cmd }(cmd)
+	defer func(env []string) { b.Config.Env = env }(env)
+
+	// derive the net build-time environment for this run. We let config
+	// environment override the build time environment.
+	// This means that we take the b.BuildEnv list of env vars and remove
+	// any of those variables that are defined as part of the container. In other
+	// words, anything in b.Config.Env. What's left is the list of build-time env
+	// vars that we need to add to each RUN command - note the list could be empty.
+	//
+	// We don't persist the build time environment with container's config
+	// environment, but just sort and pre-pend it to the command string at time
+	// of commit.
+	// This helps with tracing back the image's actual environment at the time
+	// of RUN, without leaking it to the final image. It also aids cache
+	// lookup for same image built with same build time environment.
+	cmdBuildEnv := []string{}
+	for _, bldEnv := range b.BuildEnv {
+		found := false
+		bldEnvKey := strings.Split(bldEnv, "=")[0]
+		for _, cfgEnv := range b.Config.Env {
+			cfgEnvKey := strings.Split(cfgEnv, "=")[0]
+			if bldEnvKey == cfgEnvKey {
+				found = true
+				break
+			}
+		}
+		if !found {
+			cmdBuildEnv = append(cmdBuildEnv, bldEnv)
+		}
+	}
 
-	logrus.Debugf("[BUILDER] Command to be executed: %v", b.Config.Cmd)
-
+	// derive the command to use for probeCache() and to commit in this container.
+	// Note that we only do this if there are any build-time env vars.  Also, we
+	// use the special argument "|#" at the start of the args array. This will
+	// avoid conflicts with any RUN command since commands can not
+	// start with | (vertical bar). The "#" (number of build envs) is there to
+	// help ensure proper cache matches. We don't want a RUN command
+	// that starts with "foo=abc" to be considered part of a build-time env var.
+	// A hash computed on build-env is used instead of the build env value to avoid
+	// leaking secrets into the history. For build traceability user might publish
+	// build environment as part of image commit/publish comments.
+	saveCmd := config.Cmd
+	if len(cmdBuildEnv) > 0 {
+		sort.Strings(cmdBuildEnv)
+		tmpEnv := append([]string{fmt.Sprintf("|%d", len(cmdBuildEnv))},
+			fmt.Sprintf("build env hash: %x", sha1.Sum([]byte(strings.Join(cmdBuildEnv, "")))))
```

Comment:
I wonder if we need to worry about this Join()? 
Will "a=b" and "c=d"  be confused with "a=bc=d"  (meaning, a's value is "bc=d" ) ?
Is it possible to have the []byte be "a=b\0c=d" instead?
And very minor thing (up to you), s/build env hash/env/ just to keep the line from being too long when we look at the history - in trunc mode we won't be able see as much of the key bits, which is the cmd itself.